sycl/source/detail/persistent_cache.cpp

```diff
@@ -0,0 +1,259 @@
+}
+
+/* Program binaries built for one or more devices are read from persistent
+ * cache and returned in form of vector of programs. Each binary program is
+ * stored in vector of chars.
+ */
+std::vector<std::vector<char>> PersistentCache::getPIProgramFromDisc(
+    const device &Device, const RTDeviceBinaryImage &Img,
+    const SerializedObj &SpecConsts, const std::string &BuildOptionsString,
+    RT::PiProgram &NativePrg) {
+
+  if (!isPersistentCacheEnabled())
+    return {};
+
+  // Only SPIRV images are cached
+  if (Img.getFormat() != PI_DEVICE_BINARY_TYPE_SPIRV &&
+      (Img.getFormat() == PI_DEVICE_BINARY_TYPE_NONE &&
+       pi::getBinaryImageFormat(Img.getRawData().BinaryStart, Img.getSize()) !=
+           PI_DEVICE_BINARY_TYPE_SPIRV))
+    return {};
+
+  std::string Path{
+      getCacheItemDirName(Device, Img, SpecConsts, BuildOptionsString)};
+
+  if (!OSUtil::isPathPresent(Path))
+    return {};
+
+  int i = 0;
+  std::string FileName{Path + "/" + std::to_string(i)};
+  while (OSUtil::isPathPresent(FileName + ".bin") &&
+         OSUtil::isPathPresent(FileName + ".src")) {
+    if (isCacheItemSrcEqual(FileName + ".src", Device, Img, SpecConsts,
+                            BuildOptionsString)) {
+      return readCacheItem(FileName + ".bin");
+    }
+    FileName = Path + "/" + std::to_string(++i);
+  }
+
+  return {};
+}
+
+/* Returns string value which can be used to identify different device
+ */
+std::string PersistentCache::getDeviceString(const device &Device) {
+  return Device.get_platform().get_info<sycl::info::platform::name>() + "/" +
+         Device.get_info<sycl::info::device::name>() + "/" +
+         Device.get_info<sycl::info::device::version>() + "/" +
+         Device.get_info<sycl::info::device::driver_version>();
+}
+
+/* Write built binary to persistent cache
+ * Format: numImages, 1stImageSize, Image[, NthImageSize, NthImage...]
+ */
+void PersistentCache::writeCacheItemBin(
+    const std::string &FileName, const std::vector<std::vector<char>> &Data) {
+  std::ofstream FileStream{FileName, std::ios::binary};
+
+  size_t Size = Data.size();
+  FileStream.write((char *)&Size, sizeof(Size));
+  for (size_t i = 0; i < Data.size(); ++i) {
+    Size = Data[i].size();
+    FileStream.write((char *)&Size, sizeof(Size));
+    FileStream.write(Data[i].data(), Size);
+  }
+  FileStream.close();
+}
+
+/* Read built binary to persistent cache
+ * Format: numImages, 1stImageSize, Image[, NthImageSize, NthImage...]
+ */
+std::vector<std::vector<char>>
+PersistentCache::readCacheItem(const std::string &FileName) {
+  std::vector<std::vector<char>> Res;
+  std::ifstream FileStream{FileName, std::ios::binary};
+  size_t ImgNum, ImgSize;
+  FileStream.read((char *)&ImgNum, sizeof(ImgNum));
+  Res.resize(ImgNum);
+  for (size_t i = 0; i < ImgNum; ++i) {
+    FileStream.read((char *)&ImgSize, sizeof(ImgSize));
+    Res[i].resize(ImgSize);
+    FileStream.read(Res[i].data(), ImgSize);
```

Comment:
Resize can be avoided, I believe:
```suggestion
    std::vector<char> ImgData(ImgSize);
    FileStream.read(ImgData.data(), ImgSize);
    Res[i] = std::move(ImgData);
```